Step2 - 경로 조회 기능

msmasd opened the pull request:
경로 조회 기능 구현하였습니다!
이번에도 잘 부탁드립니다~!

msmasd committed:
refac: Line에 있는 getOrderedStations 함수를 Stations로 이동 및 객체 비교시 equals로 비교

msmasd committed:
refac: 두 Section을 합치는 로직을 Section에 함수로 추출

msmasd committed:
test: 경로 인수테스트 작성

msmasd committed:
refac: 동일 기능을 하는 중복된 함수 제거

람다식으로 변경
service 필드값으로 주입받는 필드 final로 변경
RuntimeException을 NotFoundException으로 변경

msmasd committed:
feat: 최단경로조회 기능 개발

enemfk777 commented on src/main/java/nextstep/subway/line/domain/Sections.java:
```diff
@@ -51,7 +57,7 @@ public void add(Section section) {
     private boolean isStationExisted(Station station) {
         return sections.stream()
-                .anyMatch(it -> it.getUpStation() == station || it.getDownStation() == station);
+                .anyMatch(it -> Objects.equals(station, it.getUpStation()) || Objects.equals(station, it.getDownStation()));
```
section에 station을 인자로 넘겨서 직접 포함여부를 반환할 수 있을 것 같아요 :)

enemfk777 commented on src/main/java/nextstep/subway/line/domain/Sections.java:
```diff
@@ -64,31 +70,23 @@ private void raiseIfNotValidAddSection(boolean isUpStationExisted, boolean isDow
     public Optional<Section> getNextSectionByEqualUpStation(Station station) {
         return sections.stream()
-                .filter(it -> it.getUpStation() == station)
+                .filter(it -> Objects.equals(station, it.getUpStation()))
```
이 부분과 getNextSectionByEqualDownStation 역시 section이 스스로 포함여부를 반환할 수 있을 것 같습니다 :)

enemfk777 commented on src/main/java/nextstep/subway/path/domain/Path.java:
```diff
@@ -0,0 +1,51 @@
+import java.util.List;
+import java.util.Objects;
+
+public class Path {
```
Path 를 찾는 객체인데 Path라는 네이밍보단 PathFinder와 같은 네이밍은 어떠실까요? :)

enemfk777 commented on src/main/java/nextstep/subway/path/service/PathService.java:
```diff
@@ -0,0 +1,35 @@
+        Station targetStation = stationService.findStationById(target);
+
+        Path path = new Path(lineService.findAllLines());
+        GraphPath<Station, DefaultWeightedEdge> graphPath = path.findPath(sourceStation, targetStation);
```
현재의 Path는 PathFinder 등 과 같이 경로를 찾는다는 역할이 잘 드러나는 명칭으로 바꾸고,
Path라는 도메인 객체를 추가해보시는 건 어떨까요? 도메인 객체는 역들과 거리를 멤버 변수로 갖을 수 있을 것 같아요
findPath 메서드가 Path 객체를 반환하게 한다면
 Respoonse를 만들어 내는 부분도 좀 더 깔끔 해 질 수 있을 것 같습니다.

enemfk777 reviewed (changes requested):
안녕하세요 상민님 🙇‍♂️

리뷰가 늦어 대단히 죄송합니다 😭
2단계 잘 구현해 주셨네요..!
몇가지 코멘트를 드렸으니 반영 후 다시 한 번 요청 부탁드려요!!

msmasd committed:
refac: 피드백 반영

Section에 함수 추출
Path -> PathFinder로 이름 변경
PathFinder에서 Graph로 된 부분은 Path 도메인객체로 추출

msmasd committed:
refac: PathRequest 생성

enemfk777 reviewed (approved):
안녕하세요 상민님 🙇‍♂️

코멘트까지 잘 반영해주셨네요 :)
이번 단계는 이만 머지하겠습니다!
고생하셨습니다!! 💪
